### PR TITLE
CA: Use Patch to Scale clusterapi nodepools

### PR DESCRIPTION
**Pull request by Preisschild**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

Scaling up can sometimes fail in CAPI clusters with a lot of activity, due to modification conflicts (`the object has been modified; please apply your changes to the latest version and try again`). This leads to (when using the priority expander) other nodepools with a lower priority being chosen to scale up instead.

Using `.Patch()` instead of `.Update()` seemingly fixed this issue for my cluster.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/autoscaler/issues/7872

#### Special notes for your reviewer:

I'm not sure if I handled the changes for the test correctly. They pass though.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

- [client-go Scales Patch godoc]: https://pkg.go.dev/k8s.io/client-go@v0.32.3/scale#ScaleInterface
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go [hunk 1]
```diff
@@ -118,17 +121,18 @@ func (r unstructuredScalableResource) SetSize(nreplicas int) error {
...
+	if err != nil {
+		return fmt.Errorf("could not marshal json patch for scaling: %w", err)
 	}
 
-	s.Spec.Replicas = int32(nreplicas)
-	_, updateErr := r.controller.managementScaleClient.Scales(r.Namespace()).Update(context.TODO(), gvr.GroupResource(), s, metav1.UpdateOptions{})
+	_, updateErr := r.controller.managementScaleClient.Scales(r.Namespace()).Patch(context.TODO(), gvr, r.Name(), types.MergePatchType, patch, metav1.PatchOptions{})
 
 	if updateErr == nil {
 		updateErr = unstructured.SetNestedField(r.unstructured.UnstructuredContent(), int64(nreplicas), "spec", "replicas")
```
Comment (jackfrancis) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:138]: should we replace `spec` with `patch` in the error object here because that's what we're actually passing in now?

**Comment by elmiko** [comment]
makes sense to me, thank you for the fix @Preisschild !

/approve

File: cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go [hunk 2]
```diff
@@ -118,17 +121,18 @@ func (r unstructuredScalableResource) SetSize(nreplicas int) error {
 		return err
 	}
 
-	s, err := r.controller.managementScaleClient.Scales(r.Namespace()).Get(context.TODO(), gvr.GroupResource(), r.Name(), metav1.GetOptions{})
-	if err != nil {
-		return err
+	spec := autoscalingv1.Scale{
```
Comment (sbueringer) [inline cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:124]: @Preisschild @jackfrancis @elmiko Was this tested with autoscale to 0?

I was testing autoscaler v1.33 with CAPV and autoscaling to 0 is broken, while it still works with v1.32.1

It seems like the Update call from before worked but Patch does not work.

I assume because there is omitempty on autoscalingv1.ScaleSpec.Replicas

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:124]: oh, that is an interesting corner case. i did not manually test this with scale from zero. it sounds like we might want to adjust the logic here. i wonder if we could discriminate between update and patch depending on the content?

Comment (elmiko) [reply cluster-autoscaler/cloudprovider/clusterapi/clusterapi_unstructured.go:124]: fwiw, the red hat ci does test scale from/to zero, and this test does work in our infra. i will spend some time next week looking over our logs and commits to see if we are testing with this patch.
